### ISSUE TITLE
Report Unused Variable When Redefining Variable in a Child Class

### ISSUE THREAD

**Issue by diegostammer** [microsoft/pyright#1070]
**Describe the bug**
When redefining an attribute (dunder - private) in a childclass, where this attribute is NotImplemented in superclass, pyright reports reportUnusedVariable.
On superclass this dunder/private is treated as a property.
Other lintners (mypy, pylint, flake8) don't report this problem.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
class Currency():
    """
    Currency Class.

    This class represents currency.
    """

    def __init__(
        self,
        conversion_tax: float
    ) -> None:
        """Initialize the class."""
        self.__code: int = NotImplemented
        self.__abbreviation: str = NotImplemented
        self.__symbol: str = NotImplemented
        self.__name: str = NotImplemented
        self.__conversion_tax: float = conversion_tax

    @property
    def code(self) -> int:
        """Code of the money."""
        return self.__code

    @property
    def abbreviation(self) -> str:
        """Abbreviation of the money."""
        return self.__abbreviation

    @property
    def symbol(self) -> str:
        """Symbol of the money."""
        return self.__symbol

    @property
    def name(self) -> str:
        """Name of the money."""
        return self.__name


class USDollar(Currency):
    """
    USDollar Class.

    This class represents US Dollar Currency.
    """

    def __init__(self, conversion_tax: float) -> None:
        """Initialize the class."""
        super().__init__(conversion_tax)
        self.__code: int = USD
        self.__abbreviation: str = USD_ABBREVIATION
        self.__symbol: str = USD_SYMBOL
        self.__name: str = USD_NAME
```

**VS Code extension or command-line**
PyRight extension for VSCode

**Comment by erictraut** [comment]
Pyright treats double-underscore names as "private" and single-underscore names as "protected". Python doesn't have keywords for "private" and "protected" like in most other object-oriented languages. This naming convention is the typical way to distinguish between these. A "private" instance variable is not expected to be overridden by a subclass, whereas a "protected" instance variable is.

Out of curiosity, why are you using double-underscore names in this case? Did you intend to mean something other than "private"?

If you don't like this behavior, you can disable the `reportUnusedVariable` diagnostic rule on a file-by-file basis (using a `# pyright: reportUnusedVariable=false` comment) or globally using the config file.
